### ISSUE TITLE
Add mark_spent() method to Tree trait

### ISSUE THREAD

**Issue by ebfull** [zcash/incrementalmerkletree#2]
There needs to be a way to mark a witness element as "spent" so that it is later garbage collected after a sufficient number of checkpoints are registered (or something along those lines) so that it can then be discarded, to avoid accumulating large numbers of witnesses. (Currently, remove_witness cannot be called unless we're rolling back the leaf being added to the tree.)

**Comment by nuttycom** [comment]
What will probably make sense here is for each MerkleBridge to track a set of the prior leaf values that are spent in the notes that were added during the construction of that bridge, and then we can add a "garbage collection" operation that prunes the "tail" of the tree up to the checkpoint limit and discards any data that's no longer needed. This will be easier to implement after #7 merges.

**Comment by str4d** [comment]
@nuttycom notes that we don't need to make a serialization change for this, because you can implicitly track it via "things that don't have a reference". So this issue can focus on implementing the garbage collection.

**Comment by nuttycom** [comment]
Correction - there *is* a serialization change that will be needed, because we need to be able to "undo" the spends within the rollback window. So when performing garbage collection, you can only GC things that have *not* been marked as spent within the rollback window.

**Comment by nuttycom** [comment]
Marked as blocking zcash/zcash#5020 because of the required serialization change.
